server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,321 @@
+         * 5. Clear voting config
+         */
+        this.clusterState = state;
+        logger.info("initiating awareness attribute [{}] decommissioning", decommissionAttribute.toString());
+        excludeDecommissionedClusterManagerNodesFromVotingConfig(decommissionAttribute, listener);
+    }
+
+    private void excludeDecommissionedClusterManagerNodesFromVotingConfig(
+        DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        final Predicate<DiscoveryNode> shouldAbdicatePredicate = discoveryNode -> nodeHasDecommissionedAttribute(discoveryNode, decommissionAttribute);
+        List<String> clusterManagerNodesToBeDecommissioned = new ArrayList<>();
+        Iterator<DiscoveryNode> clusterManagerNodesIter = clusterState.nodes().getClusterManagerNodes().valuesIt();
+        while (clusterManagerNodesIter.hasNext()) {
+            final DiscoveryNode node = clusterManagerNodesIter.next();
+            if (shouldAbdicatePredicate.test(node)) {
+                clusterManagerNodesToBeDecommissioned.add(node.getName());
+            }
+        }
+        transportService.sendRequest(
+            transportService.getLocalNode(),
+            AddVotingConfigExclusionsAction.NAME,
+            new AddVotingConfigExclusionsRequest(clusterManagerNodesToBeDecommissioned.toArray(String[]::new)),
+            new TransportResponseHandler<AddVotingConfigExclusionsResponse>() {
+                @Override
+                public void handleResponse(AddVotingConfigExclusionsResponse response) {
+                    logger.info("successfully removed decommissioned cluster manager eligible nodes [{}] from voting config, " +
+                        "proceeding to drain the decommissioned nodes", clusterManagerNodesToBeDecommissioned.toString());
+                    registerDecommissionAttribute(decommissionAttribute, listener);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.debug(new ParameterizedMessage(
+                        "failure in removing decommissioned cluster manager eligible nodes from voting config"), exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public AddVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new AddVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    private void clearVotingConfigAfterSuccessfulDecommission() {
+        final ClearVotingConfigExclusionsRequest clearVotingConfigExclusionsRequest = new ClearVotingConfigExclusionsRequest();
+        clearVotingConfigExclusionsRequest.setWaitForRemoval(true);
```

Comment:
lets make sure that this does interfere by clearing any config exclusion that should have stayed. Is it possible to remove only those nodes that we had added earlier?

Reply:
It is not possible to do that . `clear` is also meant to be a temporary action . In case of decommission, we would clearly call out that any voting config exclusion will be cleared off. 

The other option we have is to fail decommission if any voting config exclusion is present, but it defeats the purpose of decommission .